### PR TITLE
Update/accumulo plugin 2.1

### PR DESCRIPTION
**Pull request by JWZeta**
Updated the Accumulo plugin to 2.1 and the retrieve key from server code to incorporate the Kmyth changes.

### CODE REVIEW

File: accumulo_plugin/plugin-accumulo/PelzCryptoService.java [hunk 1]
```diff
@@ -105,12 +128,13 @@ public FileEncrypter getFileEncrypter(CryptoEnvironment environment) {
 
   @Override
   public FileDecrypter getFileDecrypter(CryptoEnvironment environment) {
+    ensureInit();
     CryptoModule cm;
-    byte[] decryptionParams = environment.getDecryptionParams();
-    if (decryptionParams == null || checkNoCrypto(decryptionParams))
+    var decryptionParams = environment.getDecryptionParams();
+    if (decryptionParams.isEmpty() || checkNoCrypto(decryptionParams.get()))
       return new NoFileDecrypter();
```
Comment (b-carr) [inline accumulo_plugin/plugin-accumulo/PelzCryptoService.java:135]: Why does no encryption return a constant ```NoFileEncryptor``` (called ```DISABLED```) but no decryption directly returns ```NoFileDecryptor```?

File: accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java [hunk 2]
```diff
@@ -306,30 +379,35 @@ public void testReadNoCryptoWithCryptoConfigured() throws Exception {
...
-    CryptoService cs = clazz.getDeclaredConstructor().newInstance();
+    var cryptoProps = getAllCryptoProperties(ConfigMode.CRYPTO_TABLE_ON);
+    var droppedProperty = cryptoProps.remove(PelzCryptoService.KEY_URI_PROPERTY);
+    assertNotNull(droppedProperty);
 
-    assertEquals(PelzCryptoService.class, cs.getClass());
-    assertThrows(NullPointerException.class,
-        () -> cs.init(aconf.getAllPropertiesWithPrefix(Property.TABLE_PREFIX)));
+    String configuredClass = cryptoProps.get(INSTANCE_CRYPTO_FACTORY.getKey());
+    CryptoEnvironment env = new CryptoEnvironmentImpl(TABLE, TableId.of("5"), null);
```
Comment (b-carr) [inline accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java:398]: What does the magic 5 mean here?

File: accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java [hunk 3]
```diff
@@ -358,14 +436,38 @@ private void verifyKeySizeForCBC(SecureRandom sr, int sizeInBytes)
...
     byte[] wrapped = PelzKeyUtils.wrapKey(fek.getEncoded(), kekPath);
     assertFalse(Arrays.equals(fek.getEncoded(), wrapped));
     byte[] unwrapped = PelzKeyUtils.unwrapKey(wrapped, kekPath);
     assertTrue(Arrays.equals(unwrapped, fek.getEncoded()));
   }
+  
+  @Test
+  public void testPerTableFactory() {
+    PerTableCryptoServiceFactory factory = new PerTableCryptoServiceFactory();
+    CryptoEnvironment env = new CryptoEnvironmentImpl(TABLE, TableId.of("5"), null);
```
Comment (b-carr) [inline accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java:450]: Same question about "5".

File: include/pelz_uri_helpers.h [hunk 4]
```diff
@@ -55,7 +55,7 @@ extern "C"
  *
  * @return 0 on success, 1 on error
  */
-  int get_pelz_uri_port(UriUriA uri, int *port);
+  int get_pelz_uri_port(UriUriA uri, charbuf * port);
```
Comment (b-carr) [inline include/pelz_uri_helpers.h:58]: Why did you switch the port from an int to a charbuf?

File: sgx/pelz_enclave.edl [hunk 5]
```diff
@@ -154,18 +154,21 @@ public TableResponseStatus key_table_add_from_handle(charbuf key_id, uint64_t ha
...
  * @param[in] server_key_id.len The length of the key on server identifier
  * @param[in] key_table The key table that the new key needs to be added to
  * @param[out] key_table The key table with the new added key
  *
  * @return 0 on success, 1 on error
  */
-public TableResponseStatus key_table_add_from_server(charbuf key_id, charbuf server_id,
-  int port, charbuf server_key_id);
+public TableResponseStatus key_table_add_from_server(charbuf key_id, charbuf client_id,
+  charbuf server_id, charbuf port, charbuf server_key_id);
```
Comment (b-carr) [inline sgx/pelz_enclave.edl:171]: Why did you swap port from int to charbuf?

File: src/util/key_load.c [hunk 6]
```diff
@@ -238,11 +239,18 @@ int key_load(charbuf key_id)
         break;
       }
 
-      pelz_log(LOG_DEBUG, "Common Name: %.*s, %d", common_name.len, common_name.chars, common_name.len);
-      pelz_log(LOG_DEBUG, "Port Number: %d", port);
+      //TODO: Change this to be an input from somewhere
+      client_name = new_charbuf(10);
```
Comment (b-carr) [inline src/util/key_load.c:243]: What is ```client_name``` used for? Is it something that will come from the application that's using ```pelz```?

File: accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java (hunk 2 above)
Comment (JWZeta) [reply accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java:398]: This is a copy over from the '''CryptoTest.java''' written by the Accumulo team.

File: accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java (hunk 3 above)
Comment (JWZeta) [reply accumulo_plugin/plugin-accumulo/testfiles/PelzCryptoTest.java:450]: This is a copy over from the '''CryptoTest.java''' written by the Accumulo team.

File: include/pelz_uri_helpers.h (hunk 4 above)
Comment (JWZeta) [reply include/pelz_uri_helpers.h:58]: The port value used by enclave_retrieve_key is a char string with a length value and not an int.

File: sgx/pelz_enclave.edl (hunk 5 above)
Comment (JWZeta) [reply sgx/pelz_enclave.edl:171]: The port value used by enclave_retrieve_key is a char string with a length value and not an int.

File: src/util/key_load.c (hunk 6 above)
Comment (JWZeta) [reply src/util/key_load.c:243]: The ```enclave_retrieve_key``` function needs pelz's public certificate for the ECDH handshake with the proxy key server.  I hard coded the ```client_name``` based on how we create pelz's public certificate. I will create an issue to do this change to the pelz code.

File: accumulo_plugin/pelzPlugin/PelzObjects.java [hunk 7]
```diff
@@ -17,134 +17,116 @@
...
 
 	public static class Request {
-		private int request_type = 0;
-		private String key_id = null;
-		private int key_id_len = 0;
-		private String enc_data = null;
-		private int enc_data_len = 0;
-		private String dec_data = null;
-		private int dec_data_len = 0;
+	  private int request_type = 0;
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply accumulo_plugin/pelzPlugin/PelzObjects.java]: White-space formatting fixed

File: accumulo_plugin/pelzPlugin/PelzKeyUtils.java [hunk 8]
```diff
@@ -84,13 +84,9 @@ public static byte[] pelzRequest(PelzObjects.Request req) {
...
 	if (msg.getError() != null)
 	  throw new CryptoService.CryptoException(msg.getError());
-	if (req.getRequestType() == 1) {
-	  msg.setEncOut(msg.getEncOut().substring(0, (msg.getEncOutLen() - 1)));
-	  resp = Base64.getDecoder().decode(msg.getEncOut());
-	}
-	else if(req.getRequestType() == 2){
-	  msg.setDecOut(msg.getDecOut().substring(0, (msg.getDecOutLen() - 1)));
-	  resp = Base64.getDecoder().decode(msg.getDecOut());
+  if ((req.getRequestType() >= 1) && (req.getRequestType() <= 4)) {
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply accumulo_plugin/pelzPlugin/PelzKeyUtils.java]: The 1 thru 4 represents the current valide RequestType fields.

File: accumulo_plugin/pelzPlugin/PelzKeyUtils.java [hunk 9]
```diff
@@ -108,11 +104,10 @@ public static byte[] unwrapKey(byte[] fek, String keyId) {
...
     req.setKeyID(keyId);
-    req.setKeyIDLen(keyId.length());
-    String dec_data = Base64.getEncoder().encodeToString(fek);
-    dec_data = dec_data.concat(System.getProperty("line.separator"));
-    req.setDecData(dec_data);
-    req.setDecDataLen(dec_data.length());
+    String data = Base64.getEncoder().encodeToString(fek);
+    data = data.concat(System.getProperty("line.separator"));
+    req.setData(data);
+    req.setCipher("AES/KeyWrap/RFC3394NoPadding/256");
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply accumulo_plugin/pelzPlugin/PelzKeyUtils.java]: Changed to constant

File: README.md [hunk 10]
```diff
@@ -40,16 +40,18 @@ The JSON objects can be in two forms: requests and responses.
     * URI syntax must currently comply with RFC 8089 and RFC 1738 Section 3.1.
 * data : string of characters
     * Base64 encoded data to be processed based on request type.
+* cipher : string of characters
+    * The cipher type used to encrypt or decrypt the data
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply README.md]: added

File: README.md [hunk 11]
```diff
@@ -69,12 +71,14 @@ described above, signed requests include:
     * The key_id specified in the JSON request will be included in the JSON response.
 * data : string of characters
     * Base-64 encoded, output data based on request type.
+* cipher : string of characters
+    * The cipher type used to encrypt or decrypt the data
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply README.md]: added

**Pull request merged**